src/signing/petition_types/petition_entity.rs

```diff
@@ -0,0 +1,557 @@
+    ) -> Self {
+        Self::new(
+            intent_hash,
+            entity,
+            PetitionFactors::new_unsecurified(instance),
+            None,
+        )
+    }
+
+    /// Returns `true` signatures requirement has been fulfilled, either by
+    /// override factors or by threshold factors
+    pub fn has_signatures_requirement_been_fulfilled(&self) -> bool {
+        self.status() == PetitionFactorsStatus::Finished(PetitionFactorsStatusFinished::Success)
+    }
+
+    fn union_of<F, T>(&self, map: F) -> IndexSet<T>
+    where
+        T: Eq + std::hash::Hash + Clone,
+        F: Fn(&PetitionFactors) -> IndexSet<T>,
+    {
+        self.both(
+            |l| map(l),
+            |t, o| {
+                t.unwrap_or_default()
+                    .union(&o.unwrap_or_default())
+                    .cloned()
+                    .collect::<IndexSet<T>>()
+            },
+        )
+    }
+
+    pub fn all_factor_instances(&self) -> IndexSet<OwnedFactorInstance> {
+        self.union_of(|l| l.factor_instances())
+            .into_iter()
+            .map(|f| OwnedFactorInstance::owned_factor_instance(self.entity.clone(), f.clone()))
+            .collect::<IndexSet<_>>()
+    }
+
+    pub fn all_skipped_factor_instance(&self) -> IndexSet<HierarchicalDeterministicFactorInstance> {
+        self.union_of(|f| f.all_skipped())
+    }
+
+    pub fn all_skipped_factor_sources(&self) -> IndexSet<FactorSourceIDFromHash> {
+        self.all_skipped_factor_instance()
+            .into_iter()
+            .map(|f| f.factor_source_id)
+            .collect::<IndexSet<_>>()
+    }
+
+    pub fn all_signatures(&self) -> IndexSet<HDSignature> {
+        self.union_of(|f| f.all_signatures())
+    }
+
+    fn with_list<F, T>(list: &Option<RefCell<PetitionFactors>>, map: F) -> Option<T>
+    where
+        F: Fn(&PetitionFactors) -> T,
+    {
+        list.as_ref().map(|refcell| map(&refcell.borrow()))
+    }
+
+    fn on_list<F, R>(&self, kind: FactorListKind, r#do: &F) -> Option<R>
+    where
+        F: Fn(&PetitionFactors) -> R,
+    {
+        match kind {
+            FactorListKind::Threshold => Self::with_list(&self.threshold_factors, r#do),
+            FactorListKind::Override => Self::with_list(&self.override_factors, r#do),
+        }
+    }
+
+    fn both<F, C, T, R>(&self, r#do: F, combine: C) -> R
+    where
+        F: Fn(&PetitionFactors) -> T,
+        C: Fn(Option<T>, Option<T>) -> R,
+    {
+        let t = self.on_list(FactorListKind::Threshold, &r#do);
+        let o = self.on_list(FactorListKind::Override, &r#do);
+        combine(t, o)
+    }
+
+    fn both_void<F, R>(&self, r#do: F)
+    where
+        F: Fn(&PetitionFactors) -> R,
+    {
+        self.both(r#do, |_, _| ())
+    }
+
+    pub fn skipped_factor_source_if_relevant(&self, factor_source_id: &FactorSourceIDFromHash) {
+        self.both_void(|l| l.skip_if_references(factor_source_id, true));
+    }
+
+    /// # Panics
```

Comment:
Should it actually panic? Why not throw? And in general is there a specific reason to using panic instead of returning a result?

Reply:
it would be programmer error, I wanna know about it ASAP and fix it, because I've incorrectly used my internal types. And with 99% code coverage we know it is correctly setup.

But if we refactor and do a programmer error, then the asserts help us find them easily.